File: app/src/main/java/com/sharpdroid/registroelettronico/Fragments/FragmentLogin.java
package com.sharpdroid.registroelettronico.Fragments;

import android.content.Context;
import android.os.Bundle;
import android.preference.PreferenceManager;
import android.support.annotation.Nullable;
import android.support.design.widget.TextInputEditText;
import android.support.v7.content.res.AppCompatResources;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.WindowManager;
import android.view.inputmethod.EditorInfo;
import android.widget.Button;

import com.heinrichreimersoftware.materialintro.app.SlideFragment;
import com.orm.SugarRecord;
import com.sharpdroid.registroelettronico.API.V2.APIClient;
import com.sharpdroid.registroelettronico.Databases.Entities.LoginRequest;
import com.sharpdroid.registroelettronico.Databases.Entities.Profile;
import com.sharpdroid.registroelettronico.R;
import com.sharpdroid.registroelettronico.Utils.Account;

import butterknife.BindView;
import butterknife.ButterKnife;
import io.reactivex.android.schedulers.AndroidSchedulers;
import retrofit2.HttpException;

import static com.sharpdroid.registroelettronico.Utils.Metodi.loginFeedback;

public class FragmentLogin extends SlideFragment {

    @BindView(R.id.mail)
    TextInputEditText mEditTextMail;
    @BindView(R.id.password)
    TextInputEditText mEditTextPassword;
    @BindView(R.id.login_btn)
    Button mButtonLogin;

    private boolean loggedIn = false;
    private Context mContext;

    public FragmentLogin() {
        // Required empty public constructor
    }

    public static FragmentLogin newInstance() {
        return new FragmentLogin();
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        getActivity().getWindow().setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_ADJUST_PAN);
        mContext = getContext();
        return inflater.inflate(R.layout.fragment_login_light, container, false);
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        ButterKnife.bind(this, view);

        mEditTextMail.setCompoundDrawablesWithIntrinsicBounds(AppCompatResources.getDrawable(mContext, R.drawable.ic_person), null, null, null);
        mEditTextPassword.setCompoundDrawablesWithIntrinsicBounds(AppCompatResources.getDrawable(mContext, R.drawable.ic_password), null, null, null);

        mEditTextMail.setEnabled(!loggedIn);
        mEditTextPassword.setEnabled(!loggedIn);
        mButtonLogin.setEnabled(!loggedIn);

        mButtonLogin.setOnClickListener(v -> login());
        mEditTextPassword.setOnEditorActionListener((textView, i, keyEvent) -> {
            if (i == EditorInfo.IME_ACTION_DONE) {
                login();
                return true;
            }
            return false;
        });
    }

    @Override
    public boolean canGoForward() {
        return loggedIn;
    }

    private void login() {
        String mEmail = mEditTextMail.getText().toString();
        String mPassword = mEditTextPassword.getText().toString();

        mEditTextMail.setEnabled(false);
        mEditTextPassword.setEnabled(false);
        mButtonLogin.setEnabled(false);
        mButtonLogin.setText(R.string.caricamento);

        APIClient.Companion.with(mContext).postLogin(new LoginRequest(mPassword, mEmail))
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(login -> {
                    SugarRecord.save(new Profile(mEmail, login.getFirstName() + " " + login.getLastName(), mPassword, "", Long.valueOf(login.getIdent().substring(1, 8)), login.getToken(), login.getExpire().getTime()));

                    Account.Companion.with(getActivity()).setUser(Long.valueOf(login.getIdent().substring(1, 8)));
                    PreferenceManager.getDefaultSharedPreferences(mContext).edit()
                            .putBoolean("first_run", false)
                            .apply();

                    mButtonLogin.setText(R.string.login_riuscito);
                    loggedIn = true;
                    updateNavigation();
                    nextSlide();
                }, error -> {
                    if (error instanceof HttpException) {
                        Log.e("FragmentLogin", ((HttpException) error).response().errorBody().string());
                    }

                    loginFeedback(error, getContext());

                    mButtonLogin.setText(R.string.login);
                    mEditTextMail.setEnabled(true);
                    mEditTextPassword.setEnabled(true);
                    mButtonLogin.setEnabled(true);
                });
    }

}

File: app/src/main/java/com/sharpdroid/registroelettronico/fragments/FragmentLogin.kt
package com.sharpdroid.registroelettronico.fragments

import android.annotation.SuppressLint
import android.content.Context
import android.graphics.Color
import android.os.Bundle
import android.preference.PreferenceManager
import android.support.v7.content.res.AppCompatResources
import android.support.v7.widget.LinearLayoutManager
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.WindowManager
import android.view.inputmethod.EditorInfo
import com.afollestad.materialdialogs.MaterialDialog
import com.crashlytics.android.answers.Answers
import com.crashlytics.android.answers.LoginEvent
import com.heinrichreimersoftware.materialintro.app.SlideFragment
import com.sharpdroid.registroelettronico.BuildConfig
import com.sharpdroid.registroelettronico.R
import com.sharpdroid.registroelettronico.adapters.LoginAdapter
import com.sharpdroid.registroelettronico.api.spaggiari.v2.Spaggiari
import com.sharpdroid.registroelettronico.database.entities.LoginRequest
import com.sharpdroid.registroelettronico.database.entities.LoginResponse
import com.sharpdroid.registroelettronico.database.entities.Profile
import com.sharpdroid.registroelettronico.database.room.DatabaseHelper
import com.sharpdroid.registroelettronico.utils.Account
import com.sharpdroid.registroelettronico.utils.Metodi.fetchDataOfUser
import com.sharpdroid.registroelettronico.utils.Metodi.loginFeedback
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers
import kotlinx.android.synthetic.main.fragment_login.*
import retrofit2.HttpException
import java.util.*

class FragmentLogin : SlideFragment() {
    private var loggedIn = false
    private var mContext: Context? = null

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        activity.window.setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_ADJUST_PAN)
        mContext = context
        return inflater.inflate(R.layout.fragment_login_light, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        mail.setCompoundDrawablesWithIntrinsicBounds(AppCompatResources.getDrawable(mContext!!, R.drawable.ic_person), null, null, null)
        password.setCompoundDrawablesWithIntrinsicBounds(AppCompatResources.getDrawable(mContext!!, R.drawable.ic_password), null, null, null)

        mail.isEnabled = !loggedIn
        password.isEnabled = !loggedIn
        login_btn.isEnabled = !loggedIn

        login_btn.setOnClickListener { login() }
        password.setOnEditorActionListener { _, i, _ ->
            if (i == EditorInfo.IME_ACTION_DONE) {
                login()
                return@setOnEditorActionListener true
            }
            false
        }
    }

    override fun canGoForward() = loggedIn

    @SuppressLint("CheckResult")
    private fun login() {
        val mEmail = mail.text.toString()
        val mPassword = password.text.toString()

        mail.isEnabled = false
        password.isEnabled = false
        login_btn.isEnabled = false
        login_btn.setText(R.string.caricamento)

        Spaggiari(null).api().postLogin(LoginRequest(mPassword, mEmail, ""))
                .observeOn(AndroidSchedulers.mainThread()).subscribe({ login ->
            val checkedIdents = ArrayList<String>()
            if (login.choices != null) {
                val builder = MaterialDialog.Builder(context).title("Account multiplo").content("Seleziona gli account che vuoi importare").positiveText("OK").neutralText("Annulla")
                        .alwaysCallMultiChoiceCallback()
                        .dividerColor(Color.TRANSPARENT)
                        .adapter(LoginAdapter(login.choices, context) { checked ->
                            checkedIdents.clear()
                            checkedIdents.addAll(checked)
                            Unit
                        }, LinearLayoutManager(context, LinearLayoutManager.VERTICAL, false))
                        .onPositive { _, _ ->
                            loginWithIdent(mEmail, mPassword, checkedIdents.toTypedArray())
                                    .observeOn(AndroidSchedulers.mainThread())
                                    .subscribe({ t ->
                                        val profile = Profile(mEmail, t.firstName + " " + t.lastName, mPassword, "", t.ident!!.substring(1, 8).toLong(), t.token!!, t.expire!!, t.ident, true)
                                        DatabaseHelper.database.profilesDao().insert(profile)
                                        fetchDataOfUser(profile)
                                    }, {
                                        loginFeedback(it, context)

                                        login_btn.setText(R.string.login)
                                        mail.isEnabled = true
                                        this.password.isEnabled = true
                                        login_btn.isEnabled = true
                                    }, {
                                        PreferenceManager.getDefaultSharedPreferences(context).edit().putBoolean("first_run", false).apply()
                                        login_btn.setText(R.string.login_riuscito)

                                        postLogin()
                                    })

                        }
                        .canceledOnTouchOutside(false)
                        .onNeutral { _, _ ->
                            login_btn.setText(R.string.login)
                            mail.isEnabled = true
                            password.isEnabled = true
                            login_btn.isEnabled = true
                        }
                builder.show()
            } else {
                DatabaseHelper.database.profilesDao().insert(Profile(mEmail, login.firstName + " " + login.lastName, mPassword, "", java.lang.Long.valueOf(login.ident!!.substring(1, 8)), login.token!!, login.expire!!, login.ident, false))

                Account.with(activity).user = java.lang.Long.valueOf(login.ident.substring(1, 8))
                fetchDataOfUser(activity)
                PreferenceManager.getDefaultSharedPreferences(mContext).edit()
                        .putBoolean("first_run", false)
                        .apply()

                login_btn.setText(R.string.login_riuscito)
                postLogin()
            }
        }) { error ->
            if (error is HttpException) {
                Log.e("FragmentLogin", error.response().errorBody()?.string())
            }

            loginFeedback(error, context)

            login_btn.setText(R.string.login)
            mail.isEnabled = true
            password.isEnabled = true
            login_btn.isEnabled = true
        }
    }

    private fun postLogin() {
        loggedIn = true
        updateNavigation()
        nextSlide()
        if (!BuildConfig.DEBUG)
            Answers.getInstance().logLogin(LoginEvent().putMethod("multiple"))
    }

    private fun loginWithIdent(email: String, password: String, idents: Array<String>): io.reactivex.Observable<LoginResponse> {
        // This asterisk is called spread operator. It converts an Array to vararg
        return io.reactivex.Observable.fromArray(*idents.map { id -> Spaggiari(null).api().postLogin(LoginRequest(password, email, id)) }.toTypedArray())
                .flatMap { t -> t.subscribeOn(Schedulers.computation()) }
                .observeOn(AndroidSchedulers.mainThread())
    }
}